dask/array/tests/test_routines.py

```diff
@@ -1286,3 +1286,121 @@ def test_result_type():
     # dask 0d-arrays are NOT treated like scalars
     c = da.from_array(np.ones((), np.float64), chunks=())
     assert da.result_type(a, c) == np.float64
+
+
+@pytest.mark.parametrize('einsum_signature', [
+    'abc,bad->abcd',
+    'abcdef,bcdfg->abcdeg',
+    'ea,fb,abcd,gc,hd->efgh',
+    'ab,b',
+    'aa',
+    'a,a->',
+    'a,a->a',
+    'a,a',
+    'a,b',
+    'a,b,c',
+    'a',
+    'ba,b',
+    'ba,b->',
+    'defab,fedbc->defac',
+    'ab...,bc...->ac...',
+    'a...a',
+    'abc...->cba...',
+    '...ab->...a',
+    'a...a->a...',
+    # Following 2 from # https://stackoverflow.com/a/19203475/1611416
+    '...abc,...abcd->...d',
+    'ab...,b->ab...',
+    # https://github.com/dask/dask/pull/3412#discussion_r182413444
+    'aa->a',
+    'ab,ab,c->c',
+    'aab,bc->ac',
+    'aab,bcc->ac',
+    'fdf,cdd,ccd,afe->ae',
+    'fff,fae,bef,def->abd',
+])
+def test_einsum(einsum_signature):
+    # einsum label dimensions
+    dimensions = {'a': 5, 'b': 6, 'c': 7,
+                  'd': 5, 'e': 6, 'f': 10,
+                  'g': 1, 'h': 2, '*': 11}
+
+    # dimension chunks sizes
+    chunks = {'a': (2, 3), 'b': (2, 3, 1), 'c': (2, 3, 2),
+              'd': (4, 1), 'e': (2, 4),    'f': (1, 2, 3, 4),
+              'g': 1,      'h': (1, 1),    '*': 11}
+
+    def shape_from_string(s):
+        return tuple(dimensions[c] for c in s)
+
+    def chunks_from_string(s):
+        return tuple(chunks[c] for c in s)
+
+    input_sigs = (einsum_signature.split('->')[0]
+                                  .replace("...", "*")
+                                  .split(','))
+
+    shapes = [shape_from_string(s) for s in input_sigs]
+    chunks = [chunks_from_string(s) for s in input_sigs]
+
+    np_inputs = [np.random.random(s) for s in shapes]
+    da_inputs = [da.from_array(i, chunks=c) for i, c in zip(np_inputs, chunks)]
+
+    assert_eq(np.einsum(einsum_signature, *np_inputs),
+              da.einsum(einsum_signature, *da_inputs))
+
+
+def test_einsum_broadcasting_contraction():
```

Comment:
Nit: I'm not quite sure why you call this test "broadcasting"? Based on the example, I don't see any broadcasting -- there's no use of `'...'` and all the dimension sizes match. So maybe better to remove that part of the name (here and below).

Reply:
Ah good catch. The test cases were adapted from [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/tests/test_contract.py#L247) as suggested by https://github.com/dask/dask/pull/3412#discussion_r182413444.

I've reintroduced the multiplication terms that tested the broadcast.